alibi_detect/utils/tensorflow/_saving.py

```diff
@@ -0,0 +1,945 @@
+
+    Returns
+    -------
+    A tuple containing the model and embedding config dicts.
+    """
+
+    filepath = base_path.joinpath(path)
+
+    cfg_model = None  # type: Optional[Dict[str, Any]]
+    cfg_embed = None  # type: Optional[Dict[str, Any]]
+    if isinstance(model, UAE):
+        if isinstance(model.encoder.layers[0], TransformerEmbedding):  # if UAE contains embedding and encoder
+            # embedding
+            embed = model.encoder.layers[0]
+            cfg_embed = save_embedding_config(embed, filepath.joinpath('embedding'))
+            # preprocessing encoder
+            inputs = Input(shape=input_shape, dtype=tf.int64)
+            model.encoder.call(inputs)
+            shape_enc = (model.encoder.layers[0].output.shape[-1],)
+            layers = [InputLayer(input_shape=shape_enc)] + model.encoder.layers[1:]
+            model = tf.keras.Sequential(layers)
+            _ = model(tf.zeros((1,) + shape_enc))
+        else:  # If UAE is simply an encoder
+            model = model.encoder
+    elif isinstance(model, TransformerEmbedding):
+        cfg_embed = save_embedding_config(model, filepath.joinpath('embedding'))
+        model = None
+    elif isinstance(model, HiddenOutput):
+        model = model.model
+    elif isinstance(model, tf.keras.Model):  # This must be last as TransferEmbedding is a tf.keras.Model
+        model = model
+    else:
+        raise ValueError('Model not recognised, cannot save.')
+
+    if model is not None:
+        save_model(model, filepath=filepath, save_dir='model')
+        cfg_model = {'src': path.joinpath('model')}
+    return cfg_model, cfg_embed
+
+
+def save_model(model: tf.keras.Model,
+               filepath: Union[str, os.PathLike],
+               save_dir: Union[str, os.PathLike] = 'model',
+               save_format: Literal['tf', 'h5'] = 'h5') -> None:  # TODO - change to tf, later PR
+    """
+    Save TensorFlow model.
+
+    Parameters
+    ----------
+    model
+        tf.keras.Model or tf.keras.Sequential.
+    filepath
+        Save directory.
+    save_dir
+        Name of folder to save to within the filepath directory.
+    save_format
+        The format to save to. 'tf' to save to the newer SavedModel format, 'h5' to save to the lighter-weight
+        legacy hdf5 format.
+    """
+    # create folder to save model in
+    model_path = Path(filepath).joinpath(save_dir)
+    if not model_path.is_dir():
+        logger.warning('Directory {} does not exist and is now created.'.format(model_path))
+        model_path.mkdir(parents=True, exist_ok=True)
+
+    # save classification model
+    model_path = model_path.joinpath('model.h5') if save_format == 'h5' else model_path
+
+    if isinstance(model, tf.keras.Model) or isinstance(model, tf.keras.Sequential):
+        model.save(model_path, save_format=save_format)
+    else:
+        logger.warning('No `tf.keras.Model` or `tf.keras.Sequential` detected. No model saved.')
```

Comment:
Is this branch logically possible? Or are we expecting this to be publicly used?

Reply:
We hopefully shouldn't get to this branch, since hopefully if there are any issues picking apart the `model` (i.e. if it's a `UAE` etc) this error will be raised in the function above (`save_model_config`):

https://github.com/SeldonIO/alibi-detect/blob/fc83c707d359d4e5009b7cfb84191748a5625bdd/alibi_detect/utils/tensorflow/_saving.py#L74-L75

However, there might be edge cases I haven't thought of where we successfully get to `save_model` despite there being an issue. This would be fairly problematic wrt to saving though so I have changed from a warning to an error.